docs/proposals/cmp-parameters.md

```diff
@@ -0,0 +1,192 @@
+    
+    Writing a custom CMP server might be preferable if the parameter announcement discovery code gets too complex to be 
+    an inline shell script.
+* Easy to install
+  * Installing a simple CMP or even a CMP with a custom server should be intuitive and painless.
+* Easy to use
+  * Argo CD end-users (for example, developers) should be able to
+    1. View and set parameters in the Argo CD Application UI
+    2. See the parameters reflected in the Application manifest
+    3. Easily read/modify the generated parameters in the manifest (they should be structured in a way that's easy to read)
+* Future-proof
+  * Since the rich parameters UI is an important feature for config management tools, the parameter announcement schema 
+    should be flexible enough to announce new _types_ of parameters so the UI can customize its presentation.
+* Backwards-compatible
+  * CMPs written before this enhancement should work fine after this enhancement is released.
+  * The UI should be able to handle unknown (new) parameter types. For example, if a plugin announces a parameter of 
+    type `date`, the UI should fall back to allowing text entry. The UI can then be enhanced to provide a better input
+    mechanism in a later release.
+* Proven with a rich demonstration
+  * The initial release of this feature should include a CMP implementation of the Helm config tool. This will
+    1) Serve as a rich example for others CMP developers to mimic
+    2) Allow us to decouple the Helm config management release cycle from the Argo release cycle
+    3) Allow us to work around [this bug](https://github.com/argoproj/argo-cd/issues/7291) without including the Helm 
+       SDK in the core Argo CD code
+
+### Non-Goals
+
+* Re-implementing other config management tools as CMPs (Kustomize, Jsonnet)
+
+## Proposal
+
+### Use cases
+
+Add a list of detailed use cases this enhancement intends to take care of.
+
+## Use case 1:
+
+As an Argo CD developer, I would like to be able to build Argo CD without including the Helm SDK as a dependency.
+
+The Helm SDK includes the Kubernetes code base. That's a lot of code, and it will make builds unacceptably slow.
+
+## Use case 2:
+
+As an Argo CD user, I would like to be able to parameterize manifests built by a CMP.
+
+For example, if the Argo CD administrator has installed a CMP which applies a last-mile kustomize overlay to a Helm
+repo, I would like to be able to pass values to the Helm chart without having to manually discover those parameter names
+(in other worse, they should show up in the Application UI just like with a native Helm Application). I also shouldn't 
+have to ask my Argo CD admin to modify the CMP to accommodate the values as environment variables.
+
+### Implementation Details/Notes/Constraints 
+
+#### Terms
+
+* **Parameter announcement**: an instance of a data structure which describes an individual parameter that may be applied
+  to a specific Application.
```

Comment:
The current first class support for Helm in the UI has custom/specific interfaces, and they are different than, say, those for Kustomize. What were your thoughts on providing a general structure that will fit for all tools? Also, should we provide UI customizations that can be configured and contributed by each CMP, so they for example, can customize 

1. the ordering of their parameters (currently they are in alphabetical order)
2. custom groups with titles.  eg. instead of a.b.param1, a.b.param2 on different lines, why not show it as param1 and param2, under some logical grouping?
3. Allow different font styles for different parameters.  Eg. what if some parameters are required, so do you want the label to be bolded or underlined?
4. etc.

Basically, should we give the CMP owner the ability to customize the look and feel of the UI, based on some common configuration

Reply:
Good questions. I'll answer here and also by adding info to the proposal. Some of the questions are answered in the notes I haven't transferred into the proposal yet. 😆 

> should we give the CMP owner the ability to customize the look and feel of the UI...

Yep! I'll add details to the "Easy to use" goal. But the basic idea is that parameter announcements should be able to communicate to the UI how they should be displayed.

> ...based on some common UI configuration spec that we define

We should define _some kind_ of schema. That schema will be on a spectrum from loosely- to strictly-defined.

On one end of the spectrum, we define a data structure that allows the user to describe the UI in excruciating detail. On the other end, we accept minimal information and let the UI make decisions from there.

Here's my suggestion: a parameter announcement should accept two strings, `type` and `description`. `type` tells the UI how to interpret `description`. If `type` is "string", the UI presents a simple test input box, and there's not really any need for further `description`. For other `type`s, there will be a documented schema for the `description` field. For example, if the type is `enum`, `description` might contain a JSON list of acceptable values.

The UI would implement components for common parameter types (like string, enum, multiline text, and image). CMP developers would have to contribute UI code for new types. We could offer a way for CMPs to inject UI code for their custom types, but I'd consider that out of scope for this proposal.

> the ordering of their parameters

The UI could respect the order of the list of parameter announcements from the CMP.

> custom groups with titles

Each parameter announcement could include a `section` field. The values could be used for grouping and for group headers.

> Allow different font styles for different parameters

That info could be included in the `description` field. To accommodate this, we'd probably want to define an initial `description` schema for the `text` type of "empty string or JSON object". That would be flexible enough to add features like formatting.

> what if some parameters are required

I'm not aware of any required parameters for the existing config management tools. But if we decided this was important, we could add a `required` field to the parameter announcement spec.

Reply:
@keithchong updated with some examples that should clarify.

Reply:
> We should define some kind of schema. That schema will be on a spectrum from loosely- to strictly-defined.

Were you thinking this should be part of the plugin.yaml or a new file that the CMP needs to provide? I was trying to come up with the pros and cons. Should UI-specific details be separated out?

I do like the idea of having some core, base types, like what you described.  I know there is an issue with arrays not being shown properly in the UI but that could be improved upon.  (eg. show +/- to add or remove elements)

> That info could be included in the description field. 

I edited my comment after I posted it. We should also keep in mind to support translation for these user-friendly strings.

> I'm not aware of any required parameters for the existing config management tools. 

I not sure, but that could be something to consider in the general case. In the UI, if the field is left blank, it implies it is optional.  How will the user know which fields are required and need to be filled in?

Reply:
> Were you thinking this should be part of the plugin.yaml or a new file that the CMP needs to provide?

The schema of the parameter announcement would be documented, and then the CMP would adhere to that schema when producing a JSON array of parameters for a given application. My parameter announcement schema proposal: https://github.com/argoproj/argo-cd/blob/3c972d4211b8d48d379d9b0fbede5329a5c3cb0f/docs/proposals/cmp-parameters.md#parameter-announcement-list-schema

> Should UI-specific details be separated out? 
> ...
> We should also keep in mind to support translation for these user-friendly strings.

I should rename the `description` field to `metadata` or something like that. I'm thinking of that field, not as a textual description of the parameter, but as a JSON object holding information about how the UI should present the parameter field.

> How will the user know which fields are required and need to be filled in?

It's a good question, but I think that's the status quo for native config management tools. One way this could be implemented is by adding UI support for this announcement:

```json
[
  {
    "name": "some-required-parameter",
    "description": "{\"required\": true}"
  }
]
```

Reply:
Hi Michael, not sure if you saw in the chat during today's contributors call, but Jesse mentioned using a Open API to define the types.

One thing I was thinking about was the user friendliness/usability of the CMP UI. One thing I noticed was the current lack of (at least, from what I saw) validation of parameter values. Is the onus on the UI to perform validation on a specific type? Take for example, the Helm parameter `replicaCount`? If the user inadvertently enters a letter instead of a number, do we want up-front validation, or, we let the user proceed and let generation happen and then see the resulting failure (like it is now)?  How about semantic validation?  For semantic validation, it should be done by the CMP. Should there be `validate` (alongside `generate` and `discover`) and maybe validate should be done before generate.  Thoughts?

Reply:
I think validation is definitely something we should be ready to support with parameterized CMPs. I'm not sure what role Open API (or maybe JSON schema, if that's different?) might play, but certainly worth looking at.

I'm not sure we should define a `validate` field yet, since that would be adding a feature that doesn't exist for config management tools. But I could imagine validation info being part of the `uiConfig` field (renamed from `metadata` field described above). Putting that info in the `uiConfig` field for now would leave room for a short-term improvement until we're ready to fully-define a validation system.

Can you give an example of what you mean by semantic validation?

Reply:
I think the Open API definition can be used to verify the validity of `uiConfig`, when the plugin author is customizing the look of the UI, at “development time”, in other words. It should be versioned.

For semantic validation, perhaps take the Helm example, where the user has defined a values.yaml file in the repo, and the parameter value for the path is correct. It’s fine syntactically. However, semantically, it’s wrong because the file is not in the default path that Helm expects.  The UI can’t validate this semantically and it’s very specific to the plugin tool itself.